src/snowflake/snowpark/_internal/ast/batch.py

```diff
@@ -21,6 +24,37 @@
 # The current AST version number (generated by the DSL).
 CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
 
+# All Snowpark AST entities (protobuf message types) which refer to other AST entities via their "id: int" field.
+REF_DESCRIPTOR_NAMES = {
+    "DataframeRef",
+    "RelationalGroupedDataframeRef",
+    "FnIdRefExpr",
+    "FnRef",
+    "IndirectTableFnIdRef",
+}
+
+
+def get_dependent_bind_ids(ast: Any) -> set[int]:
+    """Retrieve the dependent AST IDs required for this AST object."""
+    dependent_ids = set()
```

Comment:
how is potentially an infinite loop avoided here? I.e., when recursively calling get_dependent_ids(...), maybe have some visited check?

Reply:
will this case never happen because ids are unique?

Reply:
It can happen actually. Any dataframe APIs that accept another dataframe as a parameter would cause an infinite loop if the same dataframe is used.

Reply:
Sorry I misinterpreted your comment, we can never loop infinitely here because we are traversing the generated AST only (its populated fields), which is guaranteed to be finite and strictly acyclic.